metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -856,21 +867,24 @@ ApiError electLeader(String topic, int partitionId, boolean uncleanOk,
             return new ApiError(UNKNOWN_TOPIC_OR_PARTITION,
                 "No such partition as " + topic + "-" + partitionId);
         }
+        if ((electionType == ElectionType.PREFERRED && partition.hasPreferredLeader())
+            || (electionType == ElectionType.UNCLEAN && partition.hasLeader())) {
+            return new ApiError(Errors.ELECTION_NOT_NEEDED);
+        }
+
         PartitionChangeBuilder builder = new PartitionChangeBuilder(partition,
             topicId,
             partitionId,
             r -> clusterControl.unfenced(r),
-            () -> uncleanOk || configurationControl.uncleanLeaderElectionEnabledForTopic(topic));
-        builder.setAlwaysElectPreferredIfPossible(true);
+            () -> electionType == ElectionType.UNCLEAN);
+
+        builder.setAlwaysElectPreferredIfPossible(electionType == ElectionType.PREFERRED);
```

Comment:
To be honest, it wasn't clear to me why we passed the unclean flag through a function while we use the builder to specify whether preferred election is desired. It would be nice if we could pass through the `ElectionType` directly to make the pattern more consistent.